### ISSUE TITLE
Genuine Radar Emulation

### ISSUE THREAD

**Issue by erikquinn** [openscope/openscope#125]
_From @erikquinn on December 15, 2016 21:46_

Depends on n8rzz/atc#212.

Since we will already have to be dealing with sending updated aircraft position information, some small measures can be taken to more properly emulate the nuances of radar.

### Delayed Drawing
Updated positions of aircraft received from the server would not simply be displayed immediately. A specifiable "update rate" consistent with the 360 degree sweep period of the radar site would be implemented, and the displayed positions updated at the same frequency. For example, ASR9 radar antennas make a full sweep appx. every 4.5 seconds. So despite receiving maybe 50 updates from the server on where the aircraft are, we would only redraw the aircraft every 4.5 seconds, at their latest reported position.

---

### Artificial Sweep
Maybe at some point, an artificial sweep can be enabled. As with the above example, the aircraft's updates are delayed. However, with an artificial sweep, the positions would be updated as the radar antenna rotates ([briefly viewable here](https://youtu.be/-FTYnWID3DI?t=329)). This would be done independently of the sending of updated positions from the server. The server just sends the updated info when it's available, and all the rest is done client-side.

### Transponder-Based Identification
Despite giving lectures on primary and secondary radar a few hundred times, I'll skip that here, because there are plenty of available resources for that. The basic idea can be deduced from the following statement I've always liked to use: "A meteorite falling from the sky will be shown on primary radar, but not secondary." Primary will show *anything* that has a return echo. Secondary is quite different: the transponders within aircraft (with mode c *on*) are broadcasting a 4-digit numerical (octal) "squawk code" in every direction, as well as their current altitude. Any radar site that receives that information is able to display it.

The squawk code is one of the following (more or less):
 - `1200` (VFR)
 - `7500` (hijack)
 - `7600` (radio failure)
 - `7700` (emergency)
 - `("discrete")` - this would be any non-reserved squawk code, they are always unique for each aircraft.

For our purposes, since we only have normal IFR traffic, they would all be assigned individual discrete codes. The way it *really* works is that if the radar system's computers detect a discrete squawk code that matches one listed in the flightplan repository, it associates the thing in the sky with the flightplan, from which it can pull the callsign, flightplan, filed altitude, aircraft type, etc.

_Copied from original issue: n8rzz/atc#214_

**Comment by n8rzz** [comment]
http://vstars.metacraft.com/Assets/Files/vSTARS%20Controller's%20Guide.pdf

**Comment by adamstyl** [comment]
Good thought. Some things to note though for the optional part:
Modern CWP don't work like that now days (in the European systems that I've seen, I don't know in the US). The RDP gets all the information from the radars (it's not just one radar scanning) combines the information and sends the end result to the CWP. The CWP is updated per "sections" known as mosaic and depending on the implementation, the screen is refreshed from the upper left corner to the right and then down. There is no "sweeping ray". This is too complex and IMO it will be a luxury at this point.

Please also note that the SSR for VFR flights in Europe is not `1200` but it's `7000` ;)

Cheers

**Comment by erikquinn** [comment]
@adamstyl Yep, mosaic is used in all multi-site configurations (en route facilities, as well as some large tracons), but most of our facilities are smaller, where they are using single site mode. In fact, according to my information M98 (minneapolis) is actually still using ARTS-3E scopes (the ancient ones with the bright green photoluminescent sweep). But yes, it will definitely have to be disabled if/when we offer an en route mode instead of just approach/departure. But if it's not too performance heavy, I think updating the scope in this manner for most of our facilities would be a nice touch. :smiley: 

If we could ever find the time to make it to this issue... \*_starts dreaming_\*

**Comment by adamstyl** [comment]
@erikquinn indeed that's true. So the real challenge here is architectural after all: have 2 effects that can be applied interchangeably: `sweep`,` mosaic` and `none` for disabling effects and players can switch as it pleases them. But the most important is to have non continuous track updates (ie the Delayed Drawing). It's actually part of the controllers' training ;)

Cheers

**Comment by erikquinn** [comment]
@adamstyl Precisely. Only place I differ is that I don't think we'll leave it to the _user_ to decide the target drawing method; it should be specific to the "facility" that we are emulating, and based on their real-world equipment and configuration (see #47). The redrawing is a little too locked to the position updates to really work correctly though, so there's definitely some work to be done to get there.
